src/utils/miopen.cpp

```diff
@@ -0,0 +1,1719 @@
+  if (other.desc_) {
+    miopenDataType_t data_type;
+    int num_dims;
+    CHECK_MIOPEN(
+      miopenGetTensorDescriptorSize(
+        other.desc_,
+        &num_dims));
+    std::vector<int> dims(num_dims), strides(num_dims);
+    CHECK_MIOPEN(
+      miopenGetTensorDescriptor(
+        other.desc_,
+        &data_type,
+        dims.data(),
+        strides.data()));
+    set(data_type, dims, strides);
+  }
+}
+
+TensorDescriptor::TensorDescriptor(TensorDescriptor&& other)
+  : desc_{other.desc_} {
+  other.desc_ = nullptr;
+}
+
+TensorDescriptor& TensorDescriptor::operator=(TensorDescriptor other) {
+  swap(other, *this);
+  return *this;
+}
+
+void swap(TensorDescriptor& first, TensorDescriptor& second) {
+  std::swap(first.desc_, second.desc_);
+}
+
+void TensorDescriptor::reset(miopenTensorDescriptor_t desc) {
+  if (desc_) {
+    CHECK_MIOPEN(miopenDestroyTensorDescriptor(desc_));
+  }
+  desc_ = desc;
+}
+
+miopenTensorDescriptor_t TensorDescriptor::release() noexcept {
+  auto old_desc = desc_;
+  desc_ = nullptr;
+  return old_desc;
+}
+
+miopenTensorDescriptor_t TensorDescriptor::get() const noexcept {
+  return desc_;
+}
+
+TensorDescriptor::operator miopenTensorDescriptor_t() const noexcept {
+  return get();
+}
+
+void TensorDescriptor::create() {
+  if (!desc_) {
+    CHECK_MIOPEN(miopenCreateTensorDescriptor(&desc_));
+  }
+}
+
+void TensorDescriptor::set(
+  miopenDataType_t data_type,
+  std::vector<int> dims_in,
+  std::vector<int> strides_in) {
+
+  // Check that arguments are valid
+  if (dims_in.empty()) {
+    LBANN_ERROR("attempted to set MIOpen tensor descriptor with no dimensions");
+  }
+  if (!strides_in.empty() && dims_in.size() != strides_in.size()) {
+    LBANN_ERROR(
+      "attempted to set MIOpen tensor descriptor ",
+      "with mismatched dimensions (", dims_in.size(), ") ",
+      "and strides (",strides_in.size(),")");
+  }
+
+  std::vector<int> dims=std::move(dims_in), strides=std::move(strides_in);
+  if (dims.size() < 4)
+  {
+    switch (dims.size())
+    {
+    case 2:
+      dims = {dims[0], 1, dims[1], 1};
+      strides = {};
+      break;
+    case 3:
+      dims = {dims[0], 1, dims[1], dims[2]};
+      strides = {};
+      break;
+    default:
+      LBANN_ERROR("Dims of size 1. Don't know what to do.");
+      break;
+    }
+  }
+
+  // Assume data is contiguous if no strides are provided
+  // Note (trb 12/29/2020): MIOpen only accepts contiguous strides.
+  if (strides.empty()) {
+    strides.resize(dims.size(), 1);
+    for (int i=strides.size()-1; i>0; --i) {
+      strides[i-1] = strides[i] * dims[i];
+    }
+  }
+
+  // Set MIOpen object
+  create();
+  CHECK_MIOPEN(
+    miopenSetTensorDescriptor(
+      desc_,
+      data_type,
+      dims.size(),
+      dims.data(),
+      strides.data()));
+
+}
+
+// -----------------------------
+// FilterDescriptor
+// -----------------------------
+
+FilterDescriptor::FilterDescriptor(miopenTensorDescriptor_t desc)
```

Comment:
We had at one point discussed whether we could have the MIOpen impl of a FilterDescriptor just hold a TensorDescriptor internally instead of reimplementing a lot of the same functionality with the same primitives under a different name. Alternatively, could we get away with `using FilterDescriptor = TensorDescriptor` in this case?

Reply:
Added in 87b1b86d3a1a0e88f01e7f7d64da6e4fe31f98fd